src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -1338,43 +1338,36 @@ private BoundExpression GetDefaultParameterValue(SyntaxNode syntax, ParameterSym
 
-                            if (memberDecl != null)
+                            memberName = null;
+                            switch (memberDecl)
                             {
-                                BaseFieldDeclarationSyntax? fieldDecl;
-
-                                if (memberDecl.Kind() == SyntaxKind.PropertyDeclaration)
-                                {
-                                    var propDecl = (PropertyDeclarationSyntax)memberDecl;
-                                    EqualsValueClauseSyntax? initializer = propDecl.Initializer;
+                                case PropertyDeclarationSyntax propDecl when initializerInSpan(propDecl.Initializer, syntax):
+                                    memberName = propDecl.Identifier.ValueText;
+                                    break;
 
-                                    if (initializer != null && initializer.Span.Contains(syntax.Span))
-                                    {
-                                        memberName = propDecl.Identifier.ValueText;
-                                        break;
-                                    }
-                                }
-                                else if ((fieldDecl = memberDecl as BaseFieldDeclarationSyntax) != null)
-                                {
-                                    memberName = null;
+                                case DataPropertyDeclarationSyntax propDecl when initializerInSpan(propDecl.Initializer, syntax):
+                                    memberName = propDecl.Identifier.ValueText;
+                                    break;
 
-                                    foreach (VariableDeclaratorSyntax varDecl in fieldDecl.Declaration.Variables)
+                                case BaseFieldDeclarationSyntax fieldDecl:
+                                    foreach (var varDecl in fieldDecl.Declaration.Variables)
                                     {
-                                        EqualsValueClauseSyntax? initializer = varDecl.Initializer;
-
-                                        if (initializer != null && initializer.Span.Contains(syntax.Span))
+                                        if (initializerInSpan(varDecl.Initializer, syntax))
                                         {
                                             memberName = varDecl.Identifier.ValueText;
                                             break;
                                         }
                                     }
+                                    break;
 
-                                    if (memberName != null)
-                                    {
-                                        break;
-                                    }
-                                }
+                                    static bool initializerInSpan(EqualsValueClauseSyntax? initializer, SyntaxNode syntax)
```

Comment:
nit: we try to group local functions at the end of methods where possible to avoid interrupting flow of logic.